packages/financial-templates-lib/src/clients/OptimisticOracleClient.js

```diff
@@ -73,54 +100,125 @@ class OptimisticOracleClient {
+      return hasProposal === undefined;
+    });
+    this.unproposedPriceRequests = unproposedPriceRequests.map(event => {
+      return {
+        requester: event.returnValues.requester,
+        identifier: this.hexToUtf8(event.returnValues.identifier),
+        timestamp: event.returnValues.timestamp,
+        currency: event.returnValues.currency,
+        reward: event.returnValues.reward,
+        finalFee: event.returnValues.finalFee
+      };
+    });
+
+    // Store proposals that have NOT been disputed:
+    let undisputedProposals = proposalEvents.filter(event => {
+      const key = this._getPriceRequestKey(event);
+      const hasDispute = disputeEvents.find(disputeEvent => this._getPriceRequestKey(disputeEvent) === key);
+      return hasDispute === undefined;
+    });
+    undisputedProposals = undisputedProposals.map(event => {
+      return {
+        requester: event.returnValues.requester,
+        proposer: event.returnValues.proposer,
+        identifier: this.hexToUtf8(event.returnValues.identifier),
+        timestamp: event.returnValues.timestamp,
+        proposedPrice: event.returnValues.proposedPrice,
+        expirationTimestamp: event.returnValues.expirationTimestamp
+      };
+    });
+
+    // Filter proposals based on their expiration timestamp:
+    const isExpired = proposal => {
+      return Number(proposal.expirationTimestamp) <= Number(currentTime);
+    };
+    this.expiredProposals = undisputedProposals.filter(proposal => isExpired(proposal));
+    this.undisputedProposals = undisputedProposals.filter(proposal => !isExpired(proposal));
+
+    // Store disputes that can be settled:
+    let resolvedPrices = await Promise.all(
+      disputeEvents.map(async disputeEvent => {
+        try {
+          // When someone disputes an OO proposal, the OO requests a price to the DVM with a re-formatted
+          // ancillary data packet that includes the original requester's information:
+          const stampedAncillaryData = await this.oracle.methods
+            .stampAncillaryData(
+              disputeEvent.returnValues.ancillaryData ? disputeEvent.returnValues.ancillaryData : "0x",
+              disputeEvent.returnValues.requester
             )
-            .call(),
-        {
-          concurrency: WEB3_CALLS_BATCH_SIZE
+            .call();
+
+          // getPrice will return null or revert if there is no price available,
+          // in which case we'll ignore this dispute.
+          let resolvedPrice = revertWrapper(
+            await this.voting.methods
+              .getPrice(disputeEvent.returnValues.identifier, disputeEvent.returnValues.timestamp, stampedAncillaryData)
+              .call({
+                from: this.oracle.options.address
+              })
+          );
+          if (resolvedPrice !== null) {
+            return disputeEvent;
+          } else {
+            return null;
+          }
+        } catch (error) {
+          return null;
```

Comment:
do you want to handle this error a bit more gracefully than just returning `null`?

Reply:
I refactored this a bit to make it more concise, I agree returning `null` is a bit ugly so instead I do nothing (i.e. I don't return the `disputeEvent`) and outside of the `Promise.all(array.map)` I remove undefined entries with a one liner.